### ISSUE TITLE
rate constant adjustment for species density(molec/cm^3)

### ISSUE THREAD

**Issue by stacywalters** [AndrewJConley/MICM_chemistry#3]
The jac function in kinetics_module.F90 multiplies each non-zero element of the jacobian, jac,  by the appropriate air density term.  The multiplier is either 1, number_density_air, or number_density_air_squared.  Similarly, the subroutine compute_rates uses, where appropriate, number_density_air or number_density_air_squared in forming the rates vector.

This entails a lot of extra multiplication operations that can and I believe should be avoided for computational efficiency.  The simple solution is to move the number_density_air and number_density_air_squared multiplication into the rateConst_update subroutine.

For example the expression for rateConstants(1) would become:

rateConstants(1) = k_rateConst(1)*c_m

and the expression for rateConstants(4) would become:

rateConstants(4) = k_rateConst(4)*c_m*c_m

Then the number_density_air and number_density_air_squared terms need to be removed
from the jac function and compute_rates subroutine.

And a small point for code simplification.  You only need one associate ... end associate construct in a given routine.  As an example the sequence:

associate( c => a )
associate( d => b )
.
.
end associate
end associate

can be replaced with:

associate( c => a, d => b )
.
.
end associate

**Comment by AndrewJConley** [comment]
I agree with you, ... except ...  The scientists want the rate constant
diagnostics to _not_ contain the air density ("M") terms.

I don't have a design for this at the moment, but I expect I'll output the
rate_constants as a diagnostic, and I'll output a separate
rate_constants_normalized_by_m variable for downstream computation.  In the
meantime, I'm living with the problem.

-- Andrew

**Comment by stacywalters** [comment]
Andrew,

I'm fairly sure the "scientists" are talking about diagnostic rate
constants.
They don't care about what goes on under the hood.

Stacy
